llvm/lib/Target/X86/X86WinEHState.cpp

```diff
@@ -653,7 +672,15 @@ void WinEHStatePass::addStateStores(Function &F, WinEHFuncInfo &FuncInfo) {
       InitialState = FinalState = ParentBaseState;
     for (Instruction &I : *BB) {
       auto *Call = dyn_cast<CallBase>(&I);
-      if (!Call || !isStateStoreNeeded(Personality, *Call))
+      if (!Call)
+        continue;
+
+      if (isSehScopeEnd(*Call)) {
```

Comment:
SEH scope ends are effectively non-throwing no-op pseudo instructions, so there should be no functional difference between inserting the store before or after the no-op. For simplicity, can we insert the state stores before the scope ends? That would make this a bit more uniform.

I haven't played around with the /EHa SEH scope codegen patterns since they were added, but it would be pretty reasonable for codegen to lower them to a direct branch, which after branch folding would get optimized out anyway.

Reply:
I added the state store before the scope end. To do that, I added a condition to `getStateForCall` as unlike other invoke instructions, that reference the state of the unwind destination, this one should reference the state of the normal destination (as the unwind destination is the same as for SEH scope begin).

I also thought about adapting `calculateStateNumbersForInvokes` within `WinEHPrepare.cpp`, which fills the `InvokeStateMap`. However, I wasn't sure what is better, so I put the SEH scope end check into `getStateForCall`, as that would be the smallest change.